Fix/citizen spawn at townhall

Raycoms opened the pull request:
Fixes #350 #292 #350 #355

Fixes some navigation problems.
An old bug with the builder crashing.
And citizens now correctly spawn at their townhalls.
Also: Miner correctly mines underground.

Some work to fix some bugs/crashes, sonar + anti griefing.

Monsters can only be spawned by the colony officer +

Explosions have a seperate config toggle.
Add config option


Only friendly players can pickUp, only officer can kill citizens.

Turn on/off griefing prot with config + fix playerContainerEvent

And add new fisherman, farmer and townhall schematics.

Reset style when switching back to huts.
Adds a new command to add Officers per command

That way admins can access colonies of players.
Create file for addOfficerCommand

Fix for crash of farmer building on destroyed.



Review please

Raycoms committed:
Merge fix from 1.11

Raycoms committed:
fix merge problems.

Raycoms committed:
Should fix problems with build tool placement.

Raycoms committed:
Fix crash on loading.

Raycoms committed:
Part 1 of miner finding nodes underground correctly.

Raycoms committed:
Should fix the miner underground position.

Raycoms committed:
get the builder hotfix for the strange heights going again

Raycoms committed:
get the builder hotfix for the strange heights going again

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -567,25 +567,47 @@ private BlockPos getPositionInDirection(final EnumFacing facing, final int dista
+    @Nullable
+    private BlockPos getFloor(@NotNull BlockPos position, int depth)
+    {
+        if (depth > 50)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Assign this magic number 50 to a well-named constant, and use the constant instead. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

MinecoloniesSonar commented:
SonarQube analysis reported 7 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 3 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ShowColonyInfoCommand.java#L53](https://github.com/Minecolonies/minecolonies/blob/79e2aa548486aa7438374c33be2f17758db52f4b/src/main/java/com/minecolonies/coremod/commands/ShowColonyInfoCommand.java#L53): The Cyclomatic Complexity of this method "execute" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -567,25 +567,47 @@ private BlockPos getPositionInDirection(final EnumFacing facing, final int dista
     }
 
     /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -567,25 +567,47 @@ private BlockPos getPositionInDirection(final EnumFacing facing, final int dista
+        return floor;
+    }
+
+    /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java:
```diff
@@ -201,6 +203,16 @@ private BlockPos getMinerProxy(final BlockPos target, final double distanceToPat
             {
+                if(level.getRandomNode().getParent() != null)
+                {
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(level.getRandomNode().getParent());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'com.minecolonies.coremod.entity.ai.citizen.miner.Node' due to existing import 'com.minecolonies.coremod.entity.ai.citizen.miner.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java:
```diff
@@ -217,6 +229,23 @@ else if (targetY <= levelDepth && workerY > levelDepth)
+
+                if(level.getRandomNode().getParent() != null)
+                {
+                    List<BlockPos> nodesToTarget = new ArrayList<>();
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'nodesToTarget' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java:
```diff
@@ -217,6 +229,23 @@ else if (targetY <= levelDepth && workerY > levelDepth)
+                if(level.getRandomNode().getParent() != null)
+                {
+                    List<BlockPos> nodesToTarget = new ArrayList<>();
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(level.getRandomNode().getParent());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'com.minecolonies.coremod.entity.ai.citizen.miner.Node' due to existing import 'com.minecolonies.coremod.entity.ai.citizen.miner.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/blocks/BlockHutTownHall.java:
```diff
@@ -44,8 +45,12 @@ public void onBlockPlacedBy(
-            if (colony == null && placer instanceof EntityPlayer)
+
+            if ((colony == null
+                    || BlockPosUtil.getDistance2D(colony.getCenter(), pos) >= Configurations.workingRangeTownHall * 2 + Configurations.townHallPadding)
```
So we create a second colony when someone places the townhall somewhere else?

Raycoms replied:
no, he never reaches this spot if he tries to place his townhall outside of his own townhall range.

Raycoms replied:
He only reaches this if the placement is valid. But we can't be sure whether he is placing it again or the first time. (Inside the correct range)

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/network/messages/BuildToolPlaceMessage.java:
```diff
@@ -142,18 +145,21 @@ private static void handleHut(
+        final Block block = Block.getBlockFromName(Constants.MOD_ID + ":blockHut" + hut);
         final Colony tempColony = ColonyManager.getClosestColony(world, buildPos);
-        if (tempColony != null && !tempColony.getPermissions().hasPermission(player, Permissions.Action.MANAGE_HUTS))
+        if (tempColony != null
```
Can we flip all these & &  to || that would look way better

Raycoms replied:
It would look better I also thought about that. But we would lose the context. It would be harder to interprete.

marvin-bitterlich replied:
Hmmm maybe not worth it then...

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -910,7 +932,7 @@ else if (entity instanceof EntityMinecart)
     private BlockPos getNodeMiningPosition(BlockPos blockToMine)
     {
-        if(((BuildingMiner) getOwnBuilding()).getCurrentLevel() == null || ((BuildingMiner) getOwnBuilding()).getCurrentLevel().getRandomNode() != null)
+        if(((BuildingMiner) getOwnBuilding()).getCurrentLevel() == null || ((BuildingMiner) getOwnBuilding()).getCurrentLevel().getRandomNode() == null)
```
Can we make the casted building a variable?

Raycoms replied:
Sure

Raycoms committed:
Some work to fix some bugs/crashes, sonar + anti griefing.

Monsters can only be spawned by the colony officer +

Explosions have a seperate config toggle.
Add config option


Only friendly players can pickUp, only officer can kill citizens.


Some sonar fixes.


stage for a moment.


Turn on/off griefing prot with config + fix playerContainerEvent

And add new fisherman, farmer and townhall schematics.
Fixes #350.

Reset style when switching back to huts.
Adds a new command to add Officers per command

That way admins can access colonies of players.
Create file for addOfficerCommand


Fix for crash of farmer building on destroyed.

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -910,11 +932,16 @@ else if (entity instanceof EntityMinecart)
+        if(getOwnBuilding() instanceof BuildingMiner)
         {
-            return blockToMine;
+            BuildingMiner buildingMiner = (BuildingMiner) getOwnBuilding();
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'buildingMiner' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/pathfinding/WalkToProxy.java:
```diff
@@ -217,6 +229,23 @@ else if (targetY <= levelDepth && workerY > levelDepth)
+                if(level.getRandomNode().getParent() != null)
+                {
+                    final List<BlockPos> nodesToTarget = new ArrayList<>();
+                    com.minecolonies.coremod.entity.ai.citizen.miner.Node currentNode = level.getNode(level.getRandomNode().getParent());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Unnecessary use of fully qualified name 'com.minecolonies.coremod.entity.ai.citizen.miner.Node' due to existing import 'com.minecolonies.coremod.entity.ai.citizen.miner.*' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AUnnecessaryFullyQualifiedName)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/commands/AddOfficerCommand.java:
```diff
@@ -0,0 +1,90 @@
+            playerName = sender.getName();
+        }
+
+        Colony colony = ColonyManager.getColony(colonyId);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'colony' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/configuration/ConfigurationHandler.java:
```diff
@@ -46,6 +46,9 @@ public static synchronized void init(final File file)
               "citizenRespawnInterval", citizenRespawnInterval, CITIZEN_RESPAWN_INTERVAL_MIN, CITIZEN_RESPAWN_INTERVAL_MAX, "Citizen respawn interval in seconds");
             builderInfiniteResources = config.get(CATEGORY_GAMEPLAY, "builderInfiniteResources", builderInfiniteResources, "Does Builder have infinite resources?").getBoolean();
+            enableColonyProtection = config.get(CATEGORY_GAMEPLAY, "enableColonyProtection", enableColonyProtection, "Enable the automatic colony protection?").getBoolean();
+            turnOffExplosionsInColonies = config.get(CATEGORY_GAMEPLAY, "turnOffExplosionsInColonies", turnOffExplosionsInColonies, "Turn off explosions inside the colonies radius?").getBoolean();
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Split this 196 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented:
SonarQube analysis reported 7 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 3 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ShowColonyInfoCommand.java#L53](https://github.com/Minecolonies/minecolonies/blob/2f222dd0a639e46c46a20fd4549449df6c780ca0/src/main/java/com/minecolonies/coremod/commands/ShowColonyInfoCommand.java#L53): The Cyclomatic Complexity of this method "execute" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

Raycoms commented:
@Kostronor can you take a look at this?

Raycoms committed:
Fix exception when breaking blocks with bare hand.

Raycoms committed:
fix nullpointer exception if the template is null when creating a struct

MinecoloniesSonar commented:
SonarQube analysis reported 7 issues
* ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) 3 major
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 4 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) [ShowColonyInfoCommand.java#L53](https://github.com/Minecolonies/minecolonies/blob/65ede058d488fe41e605ec1ef62af41f927ae17d/src/main/java/com/minecolonies/coremod/commands/ShowColonyInfoCommand.java#L53): The Cyclomatic Complexity of this method "execute" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

Raycoms commented:
@Kostronor

marvin-bitterlich commented:
:shipit: 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/356/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)
